File: SimulasiAntrian.java
import javax.swing.*;
import javax.swing.border.EmptyBorder;
import java.awt.*;
import java.util.LinkedList;
import java.util.Queue;

public class SimulasiAntrian extends JFrame {

    Queue<String[]> queue = new LinkedList<>();
    int nomor = 1;

    JTextField inputNama;
    JTextArea areaAntrian;

    // Warna tema biar konsisten
    private final Color WARNA_PRIMER = new Color(41, 128, 185); // Biru
    private final Color WARNA_AKSEN = new Color(46, 204, 113); // Hijau
    private final Color WARNA_BAHAYA = new Color(231, 76, 60); // Merah
    private final Color WARNA_NETRAL = new Color(127, 140, 141); // Abu-abu
    private final Color WARNA_BG = new Color(245, 247, 250); // Abu muda

    public SimulasiAntrian() {
        setTitle("Simulasi Antrian Bank");
        setSize(500, 450);
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setLocationRelativeTo(null); // Tengah layar
        setLayout(new BorderLayout(10, 10));
        getContentPane().setBackground(WARNA_BG);

        // ===== PANEL ATAS =====
        JPanel panelAtas = new JPanel();
        panelAtas.setLayout(new BoxLayout(panelAtas, BoxLayout.Y_AXIS));
        panelAtas.setBorder(new EmptyBorder(15, 15, 10, 15));
        panelAtas.setBackground(Color.WHITE);

        JLabel label = new JLabel("Masukkan Nama Nasabah");
        label.setFont(new Font("Segoe UI", Font.BOLD, 14));
        label.setAlignmentX(Component.LEFT_ALIGNMENT);
        panelAtas.add(label);
        panelAtas.add(Box.createRigidArea(new Dimension(0, 8)));

        inputNama = new JTextField();
        inputNama.setFont(new Font("Segoe UI", Font.PLAIN, 14));
        inputNama.setMaximumSize(new Dimension(Integer.MAX_VALUE, 35));
        inputNama.setAlignmentX(Component.LEFT_ALIGNMENT);
        panelAtas.add(inputNama);
        panelAtas.add(Box.createRigidArea(new Dimension(0, 15)));

        // Panel Tombol
        JPanel panelTombol = new JPanel(new GridLayout(2, 2, 10, 10)); // Jadi 2x2 grid
        panelTombol.setBackground(Color.WHITE);
        panelTombol.setAlignmentX(Component.LEFT_ALIGNMENT);

        JButton btnAmbil = buatTombol("Ambil Antrian", WARNA_AKSEN);
        JButton btnPanggil = buatTombol("Panggil Berikutnya", WARNA_BAHAYA);
        JButton btnTampil = buatTombol("Refresh", WARNA_PRIMER);
        JButton btnReset = buatTombol("Reset Antrian", WARNA_NETRAL); // Tombol baru

        panelTombol.add(btnAmbil);
        panelTombol.add(btnPanggil);
        panelTombol.add(btnTampil);
        panelTombol.add(btnReset);

        panelAtas.add(panelTombol);
        add(panelAtas, BorderLayout.NORTH);

        // ===== AREA ANTRIAN =====
        JPanel panelTengah = new JPanel(new BorderLayout());
        panelTengah.setBorder(new EmptyBorder(0, 15, 15, 15));
        panelTengah.setBackground(WARNA_BG);

        JLabel labelDaftar = new JLabel("Daftar Antrian Saat Ini");
        labelDaftar.setFont(new Font("Segoe UI", Font.BOLD, 14));
        labelDaftar.setBorder(new EmptyBorder(0, 0, 8, 0));
        panelTengah.add(labelDaftar, BorderLayout.NORTH);

        areaAntrian = new JTextArea();
        areaAntrian.setEditable(false);
        areaAntrian.setFont(new Font("Consolas", Font.PLAIN, 14));
        areaAntrian.setMargin(new Insets(10, 10, 10, 10));
        areaAntrian.setBackground(Color.WHITE);

        JScrollPane scroll = new JScrollPane(areaAntrian);
        scroll.setBorder(BorderFactory.createLineBorder(new Color(220, 220, 220)));
        panelTengah.add(scroll, BorderLayout.CENTER);

        add(panelTengah, BorderLayout.CENTER);

        // ===== EVENT =====
        btnAmbil.addActionListener(e -> {
            String nama = inputNama.getText().trim();
            if (nama.equals("")) {
                JOptionPane.showMessageDialog(this, "Nama tidak boleh kosong!", "Peringatan", JOptionPane.WARNING_MESSAGE);
                return;
            }

            String[] data = {String.valueOf(nomor), nama};
            queue.add(data);
            nomor++;
            inputNama.setText("");
            updateTextArea();
            inputNama.requestFocus();
        });

        inputNama.addActionListener(e -> btnAmbil.doClick());

        btnTampil.addActionListener(e -> updateTextArea());

        btnPanggil.addActionListener(e -> {
            if (queue.isEmpty()) {
                JOptionPane.showMessageDialog(this, "Antrian sudah kosong", "Info", JOptionPane.INFORMATION_MESSAGE);
                return;
            }

            String[] data = queue.poll();
            String teks = "Nomor " + data[0] + " atas nama " + data[1] + ", silakan ke loket";
            JOptionPane.showMessageDialog(this, teks, "Panggilan", JOptionPane.PLAIN_MESSAGE);
            speak(teks);
            updateTextArea();
        });

        // Event tombol reset
        btnReset.addActionListener(e -> {
            int konfirmasi = JOptionPane.showConfirmDialog(
                this,
                "Yakin mau reset semua antrian? Nomor akan mulai dari 1 lagi.",
                "Konfirmasi Reset",
                JOptionPane.YES_NO_OPTION,
                JOptionPane.QUESTION_MESSAGE
            );

            if (konfirmasi == JOptionPane.YES_OPTION) {
                queue.clear();
                nomor = 1;
                updateTextArea();
                JOptionPane.showMessageDialog(this, "Antrian berhasil direset.", "Sukses", JOptionPane.INFORMATION_MESSAGE);
            }
        });

        updateTextArea();
    }

    private JButton buatTombol(String text, Color warna) {
        JButton button = new JButton(text);
        button.setFont(new Font("Segoe UI", Font.BOLD, 12));
        button.setBackground(warna);
        button.setForeground(Color.WHITE);
        button.setFocusPainted(false);
        button.setBorder(new EmptyBorder(10, 10, 10, 10));
        button.setCursor(new Cursor(Cursor.HAND_CURSOR));
        return button;
    }

    public void updateTextArea() {
        if (queue.isEmpty()) {
            areaAntrian.setText(" Belum ada antrian.\n Silakan ambil nomor terlebih dahulu.");
            return;
        }

        StringBuilder sb = new StringBuilder();
        int i = 1;
        for (String[] data : queue) {
            sb.append(String.format(" %2d. Nomor %-3s : %s\n", i++, data[0], data[1]));
        }
        areaAntrian.setText(sb.toString());
    }

    public void speak(String text) {
        new Thread(() -> {
            try {
                String safeText = text.replace("\"", "");
                ProcessBuilder pb = new ProcessBuilder("espeak", "-s", "120", "-p", "50", "-v", "id", safeText);
                Process process = pb.start();
                process.waitFor();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }).start();
    }

    public static void main(String[] args) {
        try {
            for (UIManager.LookAndFeelInfo info : UIManager.getInstalledLookAndFeels()) {
                if ("Nimbus".equals(info.getName())) {
                    UIManager.setLookAndFeel(info.getClassName());
                    break;
                }
            }
        } catch (Exception e) {}

        SwingUtilities.invokeLater(() -> {
            new SimulasiAntrian().setVisible(true);
        });
    }
}

File: SimulasiAntrian.py
import tkinter as tk
from tkinter import messagebox
from collections import deque
import pyttsx3
import threading

class SimulasiAntrian:
    def __init__(self, root):
        self.root = root
        self.root.title("Simulasi Antrian Bank")
        self.root.geometry("500x450")
        self.root.configure(bg="#F5F7FA")

        self.queue = deque()
        self.nomor = 1

        # ===== INPUT =====
        frame_atas = tk.Frame(root, bg="white", padx=15, pady=10)
        frame_atas.pack(fill="x", pady=(10, 0))

        label = tk.Label(frame_atas, text="Masukkan Nama Nasabah",
                         font=("Segoe UI", 12, "bold"), bg="white")
        label.pack(anchor="w")

        self.input_nama = tk.Entry(frame_atas, font=("Segoe UI", 12))
        self.input_nama.pack(fill="x", pady=8)
        self.input_nama.bind("<Return>", lambda e: self.ambil_antrian())

        # ===== TOMBOL =====
        frame_tombol = tk.Frame(frame_atas, bg="white")
        frame_tombol.pack(fill="x")

        btn_ambil = self.buat_tombol(frame_tombol, "Ambil Antrian", "#2ECC71", self.ambil_antrian)
        btn_panggil = self.buat_tombol(frame_tombol, "Panggil Berikutnya", "#E74C3C", self.panggil)
        btn_refresh = self.buat_tombol(frame_tombol, "Refresh", "#2980B9", self.update_text)
        btn_reset = self.buat_tombol(frame_tombol, "Reset Antrian", "#7F8C8D", self.reset)

        btn_ambil.grid(row=0, column=0, padx=5, pady=5, sticky="ew")
        btn_panggil.grid(row=0, column=1, padx=5, pady=5, sticky="ew")
        btn_refresh.grid(row=1, column=0, padx=5, pady=5, sticky="ew")
        btn_reset.grid(row=1, column=1, padx=5, pady=5, sticky="ew")

        # ===== AREA ANTRIAN =====
        frame_tengah = tk.Frame(root, bg="#F5F7FA", padx=15, pady=10)
        frame_tengah.pack(fill="both", expand=True)

        label_daftar = tk.Label(frame_tengah, text="Daftar Antrian Saat Ini",
                                font=("Segoe UI", 12, "bold"), bg="#F5F7FA")
        label_daftar.pack(anchor="w")

        self.area = tk.Text(frame_tengah, font=("Consolas", 12), height=15)
        self.area.pack(fill="both", expand=True, pady=5)
        self.area.config(state="disabled")

        self.update_text()

    def buat_tombol(self, parent, text, color, command):
        return tk.Button(parent, text=text, bg=color, fg="white",
                         font=("Segoe UI", 10, "bold"),
                         command=command, relief="flat")

    def ambil_antrian(self):
        nama = self.input_nama.get().strip()
        if not nama:
            messagebox.showwarning("Peringatan", "Nama tidak boleh kosong!")
            return

        self.queue.append((self.nomor, nama))
        self.nomor += 1
        self.input_nama.delete(0, tk.END)
        self.update_text()

    def panggil(self):
        if not self.queue:
            messagebox.showinfo("Info", "Antrian sudah kosong")
            return

        nomor, nama = self.queue.popleft()

        teks = f"Nomor {nomor}, atas nama {nama}. Silakan menuju loket."

        messagebox.showinfo("Panggilan", teks)

        threading.Thread(target=self.speak, args=(teks,)).start()
        self.update_text()

    def reset(self):
        konfirmasi = messagebox.askyesno("Konfirmasi", "Yakin reset antrian?")
        if konfirmasi:
            self.queue.clear()
            self.nomor = 1
            self.update_text()
            messagebox.showinfo("Sukses", "Antrian direset")

    def update_text(self):
        self.area.config(state="normal")
        self.area.delete(1.0, tk.END)

        if not self.queue:
            self.area.insert(tk.END, "Belum ada antrian.\nSilakan ambil nomor.")
        else:
            for i, (nomor, nama) in enumerate(self.queue, start=1):
                self.area.insert(tk.END, f"{i:2d}. Nomor {nomor:<3} : {nama}\n")

        self.area.config(state="disabled")

    def speak(self, text):
        engine = pyttsx3.init()

        engine.setProperty('rate', 120)  # lebih lambat

        # Tambahkan jeda (gunakan koma atau titik)
        text = text.replace(",", ", ").replace(".", "... ")

        engine.say(text)
        engine.runAndWait()


# ===== MAIN =====
if __name__ == "__main__":
    root = tk.Tk()
    app = SimulasiAntrian(root)
    root.mainloop()